### ISSUE TITLE
Compass problems on Sony Xperia Z1/Z2 on Android >5.0

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#4921]
Several users independently report problems with using the compass after they upgraded their device from 4.x to 5.x.
Users report that the compass is either stuck in one direction or does only rarely move.
Other apps (such as GPS Status&Toolbox and Groundspeak app) are working normal. 

Example sysinfo:
```
.--- System information ---
Device: C6903 (C6903, Sony)
Android version: 5.0.2
Android build: 14.5.A.0.242
Cgeo version: 2015.03.31
Google Play services: enabled
Low power mode: inactive
Compass capabilities: yes
Rotation sensor: present
Geomagnetic rotation sensor: present
Orientation sensor: present
Hide own/found: true
Map strategy: auto
HW-acceleration: enabled (default state)
geocaching.com: logged in (Login OK) / Premium
System language: de_DE
Installed cgeo plugins: none
--- End of system information ---
```
```
--- System information ---
Device: D6503 (D6503, Sony)
Android version: 5.0.2
Android build: 23.1.A.0.726
Cgeo version: 2015.03.31
Google Play services: enabled
Low power mode: inactive
Compass capabilities: yes
Rotation sensor: present
Geomagnetic rotation sensor: present
Orientation sensor: present
Hide own/found: false
Map strategy: auto
HW-acceleration: enabled (default state)
geocaching.com: logged in (Login OK) / Basic
System language: de_DE
Installed cgeo plugins: none
--- End of system information ---
```

I have no clear clue yet if disabling "Google Play Services" in c:geo does make any difference.
For some users it seems to help, for others not.

Additionally one user made me aware that there seems to be a generic issue (however as mentioned above other apps compass is working nontheless):
http://www.android-hilfe.de/karten-navigation/685355-c-geo-navi-problem-seit-android-5-02-a.html

**Comment by rsudev** [comment]
Comment from a user on geoclub.de (https://forum.geoclub.de/viewtopic.php?p=1219066#p1219066, translated by me):
"I have disabled Google-Play Services and then rebooted the Phone. Now the compass works. Last time I only disabled Google Play without rebooting. If it keeps working, I am satisfied."

**Comment by Megachip** [comment]
seems not a c:geo related problem. same on GPS Tools.

but AndroidTS works.

**Comment by Lineflyer** [comment]
Lot os users on support mail with Xperia and compass problems.
Can we give any advise (e.g. via Facebook)?

@rsudev 
Is it enough to disable Google Play Services and reboot ?
Any personal experiences?

**Comment by Lineflyer** [comment]
FYI:
i scheduled a posting on FB for tomorrow:

To all Sony Xperia users...please share:
We get quite some reports of problems with c:geo after upgrading your devices to Android 5.x. Those are related to firmware errors but not to c:geo.

If you experience general instability and crashes a factory reset of your mobile phone might help.

If you have the problem, that the compass is no longer working please disable "Google Play Services" in c:geo Menu- Settings - System and reboot your device.

**Comment by rsudev** [comment]
Sorry for not coming back to you.
What you wrote is in general the solution to get a working compass.
But there is a catch. It seems, that other Apps that use Google Play Services might also affect c:geo. I experienced that with cgeo-wear. After using it for navigation, the compass in c:geo is frozen, as after disabling GMS in c:geo settings, until the next reboot.

**Comment by Megachip** [comment]
Also bei mir brachte das deaktivieren der PlayServices in den c:geo Einstellungen gefolgt von einem Neustart nichts :(

**Comment by rsudev** [comment]
Translation: For me, deactivating play services in the c:geo settings and than rebooting didn't help

@Megachip I assume 'Neustart' also meant 'rebooting the phone', not only 'restarting the app'.
If this is the case, please specifiy which phone model and rom, android and c:geo version you are using (system information in c:geo about), so we can see what might be the issue for you. Please also describe what behaviour you are exactly experiencing, it might be a different issue indeed.

**Comment by Megachip** [comment]
@rsudev sorry, yeah. Wrong language ;) Rebooting the phone. 

c:geo 2015.03.31
Phone: D6503
Rom: Stock (23.1.A.1.28 and the Version before)

Compass stuck on initial Position and did not move when moving or rotating the phone.

**Comment by kadamski** [comment]
I can confirm that problem on Xperia Z Ultra and on Xperia Z1 Compact. It is not, however, 100% reproducible. With Google Play Services enabled, it sometimes works for the first time and only freezes the 2nd time the compass activity is started. On both phones, disabling Google Play Services and rebooting the phone makes the problem go away. I don't really get why but killing the c:geo application (even using "force stop" in applications settings) is not enough, full reboot is needed.

Another possible workaround is to disable "high accuracy" in location settings and use "device only" mode. Yet another one is to remove SIM card from the phone :)

**Comment by kadamski** [comment]
One more interesting observation - when this happens, the compass does work on for example GPS Status (which probably does not use Google Play Services) but does NOW work on Google Maps (the direction arrow is frozen or non existent). I guess compass won't work for any application that uses Google Play Services.

This would suggest that it's not c:geo issue at all but on the other hand I wasn't able to reproduce this issue using only Google Maps or any other application. In fact, if I launch Google Maps before launching c:geo, I can't reproduce the issue.

**Comment by rsudev** [comment]
As I can trigger this behaviour by invoking c:geo wear (which uses Google Play service), which also is a completely separate app, it seems to clearly be an interaction of
* The hardware and Android version from Sony
* Google play services
* Interaction of c:geo with the Android location Api/GMS

Currently I don't see a real chance to find a way to fix this from c:geo side.

**Comment by Lineflyer** [comment]
We have added a workaround description to the FAQ and (kind of) verified that it is not c:geo causing this.
So closing with "Wont Fix" for now.

**Comment by Andrew17856** [comment]
Hi

I'm one of the affected Lollipop users with an Xperia device and frozen compass.  I've seen the new FAQ on it, and noticed that this issue tracker has closed it as "won't fix".

I'm a developer myself, and noticed that you appear to be building using Google Play Services (location) version 7.5.  The current version (which is also on the Xperia) is 7.8.0 - see here:
https://developers.google.com/android/guides/setup.  I've had problems before with various Google Play Services functionality when building to lower versions than those on the device (which is what happens when you upgrade to Lollipop) - you might wish to try upgrading the library version.

Other compass apps (e.g. Tracky, Maverick, GPS Status & Toolbox) all work fine, so I don't think it's a firmware issue.

Also, disabling Google Play Services as suggested in the FAQs isn't really a viable work-around.  Large amounts of functionality in any Android device depends upon Google Play Services - turning it off means lots of apps won't run at all, or run with significantly reduced functionailty.

Many thanks,

Andrew

**Comment by rsudev** [comment]
@Andrew17856 'Disable Google Play Services' is a setting of c:geo, so nothing that directly affects other apps or the overall operation of the phone.
Regarding the GMS versions: I doubt that the version mismatch is driving the original issue, as IIRC the version we used to build was aligned with the version on the phone back then (I am also one of the affected users). Nevertheless I'll try to give the upgrade a spin.

**Comment by rsudev** [comment]
Did a cross-check building against GMS 7.8. Unfortunately no change...

**Comment by samueltardieu** [comment]
Google Play services and compass are independent.

**Comment by samueltardieu** [comment]
@Andrew17856 Could you please launch cgeo from scratch, trigger a compass fault, and put the output of "adb logcat -s -v time cgeo AndroidRuntime:e" somewhere?

**Comment by kumy** [comment]
My wife's Galaxy S4 has also a non-working compass.

```
--- System information ---
Device: GT-I9505 (jfltexx, samsung)
Android version: 4.3
Android build: JSS15J.I9505XXUEMJ7
Cgeo version: 2015.08.03
Google Play services: enabled
Low power mode: inactive
Compass capabilities: yes
Rotation sensor: present
Geomagnetic rotation sensor: absent
Orientation sensor: present
Hide own/found: false
Map strategy: auto
HW acceleration: enabled (default state)
System language: fr_FR
Geocaching sites enabled:
- geocaching.com: logged in (Connecté) / Premium
Installed cgeo plugins:  none
--- End of system information ---
```

**Comment by samueltardieu** [comment]
@kumy Perfect (more devices to find the problem on). Can you try to set `useRotationSensor` to `true` inconditionally in `Sensors.java` around line 105 and tell me if it changes anything?

**Comment by kumy** [comment]
@samueltardieu Ok, set the variable to `true`.

* With `GPS and magnetic Compass used`, same behaviour, ie: compass stuck/fixed in only one direction, no move.
* With `GPS only used`, compass move, but the direction isn't good. Turning the phone doesn't follow the direction.

Maybe that's a real phone's bug, I'll cross check with another Compass application.

**Comment by kumy** [comment]
Phone's compass is okay with another application.
The application asked me to calibrate the compass by turning the phone in the x/y/z axis.

I've force closed c;geo, now:
* With `GPS and magnetic Compass used`, the compass now give an approximate GOOD direction (just a little offset of some degrees)
* With `GPS only used`, compass is now totally stuck (strange).

**Comment by kumy** [comment]
`useRotationSensor = false;` gives exact same results.

**Comment by Andrew17856** [comment]
I think if "GPS only" is being used, the phone can only determine your direction if you move.  The compass then points in the direction of your movement.  So if you're standing still on "GPS only", the compass will appear fixed.

**Comment by rsudev** [comment]
Here is the requested log (not sure if it is still relevant. some personal information removed):
```
--------- beginning of system
--------- beginning of main
08-25 22:00:56.859 I/cgeo    (10382): [main] Google Play services are available
08-25 22:00:56.887 D/cgeo    (10382): [main] OrientationProvider: sensor found
08-25 22:00:56.930 I/cgeo    (10382): [looper callbacks] No last known location available, using home location
08-25 22:00:56.936 D/cgeo    (10382): [looper callbacks] GeoDataProvider: starting the GPS and network listeners
08-25 22:00:57.164 D/cgeo    (10382): [network-1] GET https://cgeo-status.herokuapp.com/api/status.json?version_code=20121203&version_name=developer+build&locale=de_DE
08-25 22:00:57.166 I/cgeo    (10382): [main] Starting cgeo.geocaching 20121203 a.k.a developer build
08-25 22:00:57.200 D/cgeo    (10382): [RxCachedThreadScheduler-2] Database clean: started
08-25 22:00:57.209 D/cgeo    (10382): [RxCachedThreadScheduler-2] Database clean: removing 0 geocaches from listId=0
08-25 22:00:57.211 D/cgeo    (10382): [looper callbacks] GpsStatusProvider: starting the GPS status listener
08-25 22:00:57.212 D/cgeo    (10382): [RxCachedThreadScheduler-2] Database clean: removing obsolete log images records
08-25 22:00:57.246 D/cgeo    (10382): [network-2] GET https://www.geocaching.com/login/default.aspx
08-25 22:00:57.262 D/cgeo    (10382): [network-3] GET http://www.opencaching.de/okapi/services/users/user.....
08-25 22:00:57.940 D/cgeo    (10382): [RxCachedThreadScheduler-2] Database clean: finished
08-25 22:00:58.621 D/cgeo    (10382): [network-3] 200 (1359 ms) GET http://www.opencaching.de/okapi/services/users/user....
08-25 22:00:59.001 D/cgeo    (10382): [network-1] 200 (1834 ms) GET https://cgeo-status.herokuapp.com/api/status.json?version_code=20121203&version_name=developer+build&locale=de_DE
08-25 22:00:59.571 D/cgeo    (10382): [network-2] 200 (2325 ms) GET https://www.geocaching.com/login/default.aspx
08-25 22:00:59.789 I/cgeo    (10382): [network-2] Already logged in Geocaching.com as ******* (Premium)
08-25 22:00:59.789 I/cgeo    (10382): [network-2] Geocaching.com language already set to English
08-25 22:00:59.791 D/cgeo    (10382): [network-2] GET http://www.geocaching.com/email/
08-25 22:01:00.279 D/cgeo    (10382): [network-2] 200 (488 ms) GET http://www.geocaching.com/email/
08-25 22:01:00.652 D/cgeo    (10382): [network-2] GET https://www.geocaching.com/account/settings/preferences
08-25 22:01:00.665 D/cgeo    (10382): [network-5] GET https://www.geocaching.com/account/settings/membership
08-25 22:01:00.665 D/cgeo    (10382): [network-4] GET https://www.geocaching.com/account/settings/homelocation
08-25 22:01:04.680 D/cgeo    (10382): [network-4] 200 (4015 ms) GET https://www.geocaching.com/account/settings/homelocation
08-25 22:01:04.695 D/cgeo    (10382): [network-5] 200 (4030 ms) GET https://www.geocaching.com/account/settings/membership
08-25 22:01:04.782 D/cgeo    (10382): [network-2] 200 (4130 ms) GET https://www.geocaching.com/account/settings/preferences
08-25 22:01:04.937 D/cgeo    (10382): [network-5] Setting member status to Premium
08-25 22:01:15.806 D/cgeo    (10382): [looper callbacks] GpsStatusProvider: stopping the GPS status listener
08-25 22:01:18.309 D/cgeo    (10382): [looper callbacks] GeoDataProvider: stopping the GPS and network listeners
08-25 22:01:23.898 I/cgeo    (10382): [main] No last known location available, using home location
08-25 22:01:24.032 D/cgeo    (10382): [main] LocationProvider: stopping location requests
08-25 22:01:26.753 I/cgeo    (10382): [main] Requesting settings backup with settings manager
08-25 22:01:26.765 D/cgeo    (10382): [main] LocationProvider: requesting most precise locations
08-25 22:01:26.787 D/cgeo    (10382): [looper callbacks] GpsStatusProvider: starting the GPS status listener
08-25 22:01:27.375 D/cgeo    (10382): [looper callbacks] GpsStatusProvider: stopping the GPS status listener
08-25 22:01:28.872 I/cgeo    (10382): [main] Starting cgeo.geocaching 20121203 a.k.a developer build
08-25 22:01:28.877 D/cgeo    (10382): [looper callbacks] GpsStatusProvider: starting the GPS status listener
08-25 22:01:31.941 D/cgeo    (10382): [looper callbacks] GpsStatusProvider: stopping the GPS status listener
08-25 22:01:32.070 D/cgeo    (10382): [looper callbacks] OrientationProvider: starting the orientation provider
08-25 22:01:32.076 I/cgeo    (10382): [looper callbacks] No last known location available, using home location
08-25 22:01:32.076 D/cgeo    (10382): [looper callbacks] GeoDataProvider: starting the GPS and network listeners
08-25 22:01:32.235 D/cgeo    (10382): [main] Loading GC261HA FloSi Memory 2010 (68) from DB
08-25 22:01:32.239 D/cgeo    (10382): [main] Loading GC340BW Geklaut in Köln (12) from DB
08-25 22:01:32.241 D/cgeo    (10382): [main] Loading OC246C Tübingen Marktplatz (45) from DB
08-25 22:01:32.243 D/cgeo    (10382): [main] Loading OCB996 Arm dran und reichlich eng (45) from DB
08-25 22:01:34.016 D/cgeo    (10382): [looper callbacks] OrientationProvider: stopping the orientation provider
08-25 22:01:34.113 D/cgeo    (10382): [network-7] GeokretyConnector.searchTrackables: wpt=OC246C
08-25 22:01:34.115 D/cgeo    (10382): [network-7] GET http://api.geokretymap.org/export2.php?wpt=OC246C
08-25 22:01:34.184 D/cgeo    (10382): [main] Loading OC246C Tübingen Marktplatz (45) from DB
08-25 22:01:34.663 D/cgeo    (10382): [network-7] 200 (548 ms) GET http://api.geokretymap.org/export2.php?wpt=OC246C
08-25 22:01:36.517 D/cgeo    (10382): [looper callbacks] GeoDataProvider: stopping the GPS and network listeners
08-25 22:01:37.464 D/cgeo    (10382): [main] destination set: D 1,0 · T 1,0 (N 48° 31,225 · E 009° 03,226)
08-25 22:01:37.473 D/cgeo    (10382): [looper callbacks] OrientationProvider: starting the orientation provider
08-25 22:01:37.477 I/cgeo    (10382): [looper callbacks] No last known location available, using home location
08-25 22:01:37.477 D/cgeo    (10382): [looper callbacks] GeoDataProvider: starting the GPS and network listeners
08-25 22:01:37.480 D/cgeo    (10382): [looper callbacks] GpsStatusProvider: starting the GPS status listener
```

**Comment by rsudev** [comment]
I now checked to change Sensors,java as suggested by @OneEyed on IRC:
```
@@ -100,11 +100,11 @@ public class Sensors {
         // Combine the magnetic direction observable with the GPS when compass is disabled or speed is high enough.
         final AtomicBoolean useDirectionFromGps = new AtomicBoolean(false);
 
         // The rotation sensor seems to be bogus on some devices. We should start with the orientation one, except when we explicitely
         // want to use the low-power geomagnetic rotation sensor or when we do not have an orientation sensor.
-        final boolean useRotationSensor = (useLowPower && RotationProvider.hasGeomagneticRotationSensor(app)) || !OrientationProvider.hasOrientationSensor(app);
+        final boolean useRotationSensor = true; //(useLowPower && RotationProvider.hasGeomagneticRotationSensor(app)) || !OrientationProvider.hasOrientationSensor(app);
         final Observable<Float> sensorDirectionObservable = useRotationSensor ? RotationProvider.create(app, useLowPower) : OrientationProvider.create(app);
         final Observable<Float> magneticDirectionObservable = sensorDirectionObservable.onErrorResumeNext(new Func1<Throwable, Observable<? extends Float>>() {
             @Override
             public Observable<? extends Float> call(final Throwable throwable) {
                 Log.e("Device orientation is not available due to sensors error, disabling compass", throwable);
```
With this change, the compass seems to behave similar with or without GMS active (low-power off).
I can try to test-drive this a little more in the next days.

**Comment by samueltardieu** [comment]
@rsudev You mean it seems to work now with this change?

**Comment by rsudev** [comment]
At the first look, yes. Therefore I would like to test some more, as sometimes this bug even in the past seemed a little erratic.

**Comment by samueltardieu** [comment]
It would mean that we may have to use the rotation sensor by default as some platforms may have removed the now-deprecated orientation sensor in recent Android versions.

**Comment by Andrew17856** [comment]
Thanks for looking at this and re-opening it - much appreciated.  I saw a request for a log earlier, but I think rsudev has kindly posted one.  If you need me to post one too I can, but it may be a day or so as I need my dev PC to do it.

**Comment by samueltardieu** [comment]
@Andrew17856 That should be ok with @rsudev log.

**Comment by samueltardieu** [comment]
Tomorrow's nightly (20150826-NB) will use the rotation sensor by default over the orientation sensor. Please test on as many devices as possible.

**Comment by Andrew17856** [comment]
The compass on the 26/8 nightly build works ok for me on my Z Ultra.  Thank you.

**Comment by Lineflyer** [comment]
In a short test I have to say that the compass on my Galaxy S4 is not working as it should anymore. The direction seems random sometimes responding to device rotation correctly but sometime jumpy or with an offset of 180°.
Will need more time to compare with older versions but on first experience there is something wrong or worse than before for me.

**Comment by Andrew17856** [comment]
I've just double checked on mine to ensure that the compass rotates smoothly and properly points north.   It does, so it seems ok on my Z Ultra.

**Comment by samueltardieu** [comment]
@Lineflyer Can you please check that you are not testing this near a large metallic mass, and that it looks OK with other software such as GPS test.

**Comment by Lineflyer** [comment]
Will do that next days for sure with different versions and devices...but only after the weekend.

**Comment by samueltardieu** [comment]
There is now a new preference in Settings/System that let you force the use of the orientation sensor when both of them are available.

**Comment by Lineflyer** [comment]
@samueltardieu :
Could you kindly summarize the recent changes to master once again?

My understanding:
If the new preference is deactivated the nightly should behave like the current release version, if activated a new approach (orientation sensor) is used.

Correct, or are there any general changes to handle the direction since last release?

**Comment by samueltardieu** [comment]
In release :

1. if low-power is activated, the geomagnetic rotation  vector sensor (compass with gyroscope turned off, very low precision and log reaction time)used is selected if present ;
- otherwise, the orientation sensor (deprecated) is used if present ;
- otherwise, the rotation vector sensor is used if present ;
- otherwise, the GPS only is used (devices without any compass capabilities).

(1) was causing a lot of reliability problems for the few people using it. (2) is causing problems on some devices such as the Xperia phones running Lollipop since it looks like the orientation sensor, which has been marked as deprecated for several major Android versions, is not working properly. (3) may be causing problems on the Galaxy S4 running KitKat and maybe Lollipop but is the recommended sensor to use.

In master :

1. the rotation vector sensor is used if present ;
- otherwise, the orientation sensor (deprecated) is used if present ;
- otherwise, the GPS only is used (devices without any compasss capabilities).

If you activate the settings "force use of orientation sensor" (which is enabled only when both the rotation vector sensor and the orientation sensor are present), then it will be used instead. 

Note that we have a patch specific to Samsung devices for the rotation vector sensor in https://github.com/cgeo/cgeo/blob/master/main/src/cgeo/geocaching/sensors/RotationProvider.java#L44-L52. That may be the cause of your problems with the new setup (if you have any) and we would need to work more on that to see if this workaround is still needed.

**Comment by samueltardieu** [comment]
However, @Lineflyer, would you mind waiting for my next commit (I'll rebuild a nightly) before you test?

**Comment by samueltardieu** [comment]
@Lineflyer The nightly has been built, you may test.

**Comment by Lineflyer** [comment]
Feel free. Can only test this evening or tomorrow evening on a limited set of devices.
I have also just learned, that I seem to have HW or calibration related issues with my specific S4 device (looking at the Samsung internal sensor information menus).

**Comment by Lineflyer** [comment]
> Note that the low-power settings now never influences the choice of the compass sensor used: we need a reliable compass, so I removed the geomagnetic rotation sensor altogether.

So "low power" functionality is now limited to the function to switch off GPS in some views but no change to any direction related things.

**Comment by samueltardieu** [comment]
> So "low power" functionality is now limited to the function to switch off GPS in some views but no change to any direction related things.

Yes (to be more exact, favor low-precision location information in some views if the precision is still "good enough", that is less than 20m).

**Comment by kumy** [comment]
What about the "compass calibration process" I got while installing a compass dedicated application? Is this something completely out of scope?
This has solved the/a "compass problem" on my Galaxy S4.

**Comment by samueltardieu** [comment]
We can add a link to such application in the "Tools" menu and an entry in the FAQ.

**Comment by kumy** [comment]
I've just tested with my HTC M8, Sasmsung S3 4G, and Sony Z3 compacthttp://forum.xda-developers.com/showthread.php?t=1778131

Before and after calibration with "GPS Essentials" I get really strange results...

![img_1766](https://cloud.githubusercontent.com/assets/176794/9580826/02fb0aa6-4ffc-11e5-82b2-3e9d5cfc34dc.JPG)
![img_1767](https://cloud.githubusercontent.com/assets/176794/9580827/02ff17e0-4ffc-11e5-9707-f411835da7fb.JPG)

None of them point to the right direction.
North is on Left, and cache on bottom...

**Comment by kumy** [comment]
On, Huawei G620‎, after calibration (before, north was at South!) North is good, but white arrow, should point West.

![img_20150831_172404](https://cloud.githubusercontent.com/assets/176794/9582430/991539f4-5005-11e5-9588-be853468bebb.jpg)

**Comment by samueltardieu** [comment]
Funny :)

Concerning the direction of the cache itself relative to the perceived device orientation, are you sure this is not due to the position imprecision? The cache is close enough that a small difference in the perceived position places it in different direction. You should try with a cache whose distance is greater.

Does your table have a magnetic structure by any chance? Can you try lifting the phone and see if it changes anything? Could you launch GPS test with the phone at the same position and see if the compass is also inaccurate?

Can you also check whether checking "use orientation sensor" in Settings/System changes anything?

**Comment by samueltardieu** [comment]
> On, Huawei G620‎, after calibration (before, north was at South!) North is good, but white arrow, should point West.

Due to the error on your position, it points onto the right direction (you are supposed to be located at the "Home" icon position and the cache at the "Star" one):

![photo_2015-08-31_19-26-26](https://cloud.githubusercontent.com/assets/44656/9585280/30fe4cb4-5016-11e5-8b80-b3f9ad430042.jpg)

**Comment by kumy** [comment]
I'll try to test again tomorrow.

**Comment by kumy** [comment]
> #5179 @kumy hasn't confirmed his tests.

@samueltardieu sorry for the delay. Just tested:
* outside
* on a wooden table
* with a "far" cache

All compass were good except the "Sony Z3 compact". Launching "GPS Essential", and do the calibration process fixed the compass. I should have tested if the calibration process works in c:geo before, but missed that part, and the compass is good at the moment. I don't know if this calibration process is something specific to "GPS Essential" or to "Android API"...
(HTC M8 has still a small drift, but I don't think it's related.)

![20150904_105134](https://cloud.githubusercontent.com/assets/176794/9680441/4ac05ff0-52f7-11e5-827b-a5cfae7cc284.jpg)

**Comment by Lineflyer** [comment]
Wait a second....
Looking at the different phones each one shows the same angle of 297°. Is this data also based on the rotation/orientation input?

Regarding compass calibration:
Is there really a special API or process on each device for that. My thinking was, that the sensors calibrate internally during operation and the user can "assist" that process by turning around all axis to give the HW a chance to get the correct north direction without any need to run special apps or functions. Am I wrong?

**Comment by Lineflyer** [comment]
@kumy:
Thanks for assisting with the testing. I have too less time at the moment.
Can you test continous rotation with each device (lay on the table and turn the device around 360° slowly expecting the compass needle always point to the same direction during the whole process)?
My quick and not reliable test on my S4 showed several "jumpings" during that rotation only with the rotation sensor but not using the orientation sensor.

**Comment by Lineflyer** [comment]
@samueltardieu 
After switching the setting for rotation/orientation sensor, does it require a restart of the app or will it be used on the next start of the compass activity without any preconditions?

**Comment by samueltardieu** [comment]
It's immediate.

**Comment by kumy** [comment]
@Lineflyer when turning the phones, they still point "their" north. Just some little lag/small jumps. Ie: its not fluid...

**Comment by kumy** [comment]
All tests I ve done with those devices were with the new setting disabled. Totally forgot about it until now :/
I ve enabled it on the Htc m8. The compas was stuck. Then I disabled the option thinking the compass will work as before, but not, it's still stuck. Force kill c:geo doesn't help. Compass in other app seems to work. I ll reboot to see if that help.



(I m on the road, no computer with me to get log).

**Comment by kumy** [comment]
What would be useful is a step by step procedure, with attended results.. So more people can test...

**Comment by kumy** [comment]
Reboot doesn t help. Compass isn t working anymore for me whatever the setting is.

**Comment by rsudev** [comment]
@kumy You could try disabling GMS in c:geo and then reboot, perhaps this works (that was the only configuration that worked for me before the change)

**Comment by Lineflyer** [comment]
@kumy: Users reported similar kind of issue only to be resolved by a new installation...The reports stopped quite a while ago (don't know if it was coincident with changes we made at that time). Strange issue.

Meanwhile I tested on several devices:
- HTC Wildfire S A510e (Android 2.3)
- Samsung Galaxy S3 (Android 4.4.4)
- Samsung Galaxy S4 (Android 5.0.1)

All of those have both sensors and the compass is working fine with both options  (despite small differences while rotating the devices). 
Only the HTC Wildfire seems to have a bad compass in general...very slow following but OK in general.

Will next test my other HTC devices as soon as batteries are charged.

**Comment by Lineflyer** [comment]
Here we go:

- HTC One mini (Android 4.2.2)

Everything OK as described above with both sensor settings.

- HTC One V (Android 4.0.3)

Compass stuck but thats normal as the device does not have compass sensors.

So for me it seems that the new default is OK....sadly I don't have any Sony Xperia device to test and there is still @kumy problem open to investigate.

**Comment by kumy** [comment]
@Lineflyer For me, the compass was "magically unlocked" on Saturday morning. Since I cannot reproduce the problem with any combinations of "GMS" and "Force use orientation sensor". Really weird...

**Comment by rsudev** [comment]
I tested with my Sony Xperia Z1, Android 5.0.2, Stock ROM (14.5.A.0.270), rooted.

With GMS and Force Orientation Sensor I had the expected Lockup (latest on the second usage) until next reboot.
Only GMS or only Force Orientation worked seemingly OK.

**Comment by samueltardieu** [comment]
@rsudev And without "force orientation sensor" (the default), everything is working on your Sony Xperia Z1, right?

The "lockup" you have is probably totally unrelated to the compass, since none of the direction sensors have anything to do with Google Play Services. However, they are often used at the same time (since to get a reliable direction towards a cache you need to get a reliable location of where you stand).

**Comment by rsudev** [comment]
@samueltardieu Yes, the default setting (GMS enabled, force orientation disabled) now works on my Z1.

What I saw during my various attempts to understand the behaviour was, that the orientation sensor seemed to no longer deliver new values (through the Android Api) once GMS had been active (to deliver location information), even if this was in a different App (c:geo wear for example). Today I tested with a Sensor-Readout app that the low-level Apis still deliver the correct values in this situation, so some bug interference/between the Api-Layer GMS uses and the orientation sensor in the Xperia Lollipop firmware (it was no problem under Kitkat).

**Comment by Lineflyer** [comment]
So I assume we can close this.
The strange behavior I initially reported for my Samsung Galaxy S4 using the rotation sensor vanished after carefully calibrating the compass. So it was likely a coincidence that it occured at the same time, or the rotation sensor values are somehow more sensible (need a better calibration).
